[ITP] User initiated cross-domain link navigations getting counted as Top Frame Redirects

charliewolfe opened the pull request:
#### aee01bb955fb9818f0fbb802cd5b5392773a33a4
<pre>
[ITP] User initiated cross-domain link navigations getting counted as Top Frame Redirects
<a href="https://bugs.webkit.org/show_bug.cgi?id=223405">https://bugs.webkit.org/show_bug.cgi?id=223405</a>
rdar://75552559

Reviewed by Wenson Hsieh and David Kilzer.

This patch fixes a use-after-move on `userInitiatedActivity`.
The variable `wasPotentiallyInitiatedByUser` uses the RefPtr
`userInitiatedActivity` to determine if a navigation was user
initiated. Since it has already been moved, the RefPtr will
always be null. This causes an issue where ITP may incorrectly
treat a user-initiated navigation as a main frame redirect.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):

Canonical link: <a href="https://commits.webkit.org/258382@main">https://commits.webkit.org/258382@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b97bd704f098ea0d00662fd25394f55e6677699

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101668 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34738 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111002 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11774 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1729 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108764 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107448 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/36568 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90883 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78557 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4416 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25166 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4491 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1615 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10579 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44654 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6245 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3032 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh reviewed (approved):
Nice catch — r=mews

ddkilzer reviewed (approved):
r=me.  It's scary that this doesn't appear to be caught by the clang static analyzer.

charliewolfe committed:
[ITP] User initiated cross-domain link navigations getting counted as Top Frame Redirects
https://bugs.webkit.org/show_bug.cgi?id=223405
rdar://75552559

Reviewed by Wenson Hsieh and David Kilzer.

This patch fixes a use-after-move on `userInitiatedActivity`.
The variable `wasPotentiallyInitiatedByUser` uses the RefPtr
`userInitiatedActivity` to determine if a navigation was user
initiated. Since it has already been moved, the RefPtr will
always be null. This causes an issue where ITP may incorrectly
treat a user-initiated navigation as a main frame redirect.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):

Canonical link: https://commits.webkit.org/258382@main
